### ISSUE TITLE
Improve the python implementation of some search algorithms

### ISSUE THREAD

**Issue by codefeeder** [aimacode/aima-python#1068]
### Issue Type
Code improvement

### Changes
Returning the complete path from the root node to the goal node from existing functions for search algorithms in search.py

### Descriptions
The current recursive and iterative algorithms for breadth-first search, depth-first search, etc. only returns the node or None when goal node is found or not found respectively.
I propose to improve the algorithm to return the complete path from the start node to the goal node in case a path is found otherwise return None.

I also propose to modify driver functions accordingly.

@MrDupin may I work on this?

**Comment by MrDupin** [comment]
This sounds good. What do you mean by 'driver functions' though?

**Comment by codefeeder** [comment]
By 'driver functions' I am referring to the functions calling the search functions as their code also would need to be modified to accommodate  the changes in the return values of the search functions.

**Comment by MrDupin** [comment]
Hmmm, if the driver function changes are only in the way they get return values, then it's fine. I don't want to modify them that much, since they must be as close to the pseudocode as possible.

**Comment by codefeeder** [comment]
@MrDupin I have made the changes and as suggested by you lso, I made minimal changes in the test_search.py file in PR #1070 
My changes pass all the test cases provided in test_search.py but during final testing the output is expected in the old format and therefore, some test cases failed.
Should I format the output according to the old requirements?

**Comment by MrDupin** [comment]
What is the old format and what is the new format of the output?

**Comment by codefeeder** [comment]
![image](https://user-images.githubusercontent.com/22821850/56082119-8c7eb280-5e32-11e9-813c-d249b27f3a8e.png)
![image](https://user-images.githubusercontent.com/22821850/56082128-a6b89080-5e32-11e9-886a-bff67a9b51ee.png)
The old format("<Node ()>") is created by using the single node returned by the search functions.
But, the newly modified search program is directly returning the list of nodes in the path without any further computation.
But due to the difference in the formats, the above test case failed even though the values were same.

**Comment by MrDupin** [comment]
Hmm, you can change the format, sure. What I'm not sure about now is the implementation itself. It seems that you are using quite a lot more lines to find the full path. Can we not use `Node`'s `parent` attribute to backtrack through the path?

**Comment by codefeeder** [comment]
Yes, sir, we can but that would increase the computation by again backtracking from the current node to the root which according to me can be done at the same time we finding a path.

**Comment by MrDupin** [comment]
The issue is that the code now gets more complicated, which is not desirable. Remember that this project is educational in nature and efficiency comes second. Operations should be done in the simplest and most intuitive manner possible.

**Comment by codefeeder** [comment]
Sure sir, I got your point. Sir, any suggestions on as to how I should proceed in order to add the functionality without adding any further complexity would be appreciated.

**Comment by MrDupin** [comment]
I am not that familiar with the search module, but can't we backtrack from node.parent to node.parent?

**Comment by dipzzzz** [comment]
@codefeeder were you able to improve your algorithm? Is this issue closed?

**Comment by dipzzzz** [comment]
As per me, create custom test cases or use regex to remove <Node ()> and then compare
Or simplest way is to add in your output

**Comment by dmeoli** [comment]
Addressed in master. The search functions deliberately return the goal `Node` (staying close to the book's pseudocode), and the full path is read off it without changing them — exactly as suggested here (backtracking via `Node.parent`):
- `node.path()` → the list of nodes from root to goal,
- `node.solution()` → the corresponding actions,
- and now `node.path_states()` → the path as a list of states (the one piece that previously needed a manual comprehension).

```python
node = astar_search(GraphProblem('Arad', 'Bucharest', romania_map))
node.path_states()   # ['Arad', 'Sibiu', 'Rimnicu', 'Pitesti', 'Bucharest']
node.solution()      # ['Sibiu', 'Rimnicu', 'Pitesti', 'Bucharest']
```

This keeps the algorithms simple (the concern that led to closing the original PR #1070) while making the complete path trivial to get. Closing as done.
